File: python/score.py
class Score:
    def __init__(self, player_1, player_2):
        self.player_1 = player_1
        self.player_2 = player_2
        self.player_in_advantage = None

    def compute_score(self):
        score_data = {}

        if self.player_1.points > self.player_2.points:
            score_data['player_in_advantage'] = self.player_1.name
            self.player_in_advantage = self.player_1.name
        else:
            score_data['player_in_advantage'] = self.player_2.name
            self.player_in_advantage = self.player_2.name

        if self.player_1.points >= 4 or self.player_2.points >= 4:
            score_data['advantage_mode'] = True
            if abs(self.player_1.points - self.player_2.points) >= 2:
                score_data['game_over'] = True

        return score_data

    def print_score(self, score_data):
        if score_data.get('advantage_mode') \
                and score_data.get('player_in_advantage'):
            return f"Advantage {score_data['player_in_advantage']}"

        return f'{self._convert_tennis_points(self.player_1.points)}-' \
               f'{self._convert_tennis_points(self.player_2.points)}'

    @staticmethod
    def _convert_tennis_points(points):
        return {
            0: "Love",
            1: "Fifteen",
            2: "Thirty",
            3: "Forty",
        }[points]


class TieScore(Score):
    def print_score(self, score_data):
        return self._convert_tie_points(self.player_1.points)

    def _convert_tie_points(self, points):
        if points < 3:
            return f"{self._convert_tennis_points(points)}-All"
        else:
            return "Deuce"


class GameOver(Score):
    def print_score(self, score_data):
        return f"Win for {score_data['player_in_advantage']}"


class AdvantageScore(Score):
    def print_score(self, score_data):
        return f"Advantage {self.player_in_advantage}"

File: typescript/src/Score.ts
abstract class Score {
    protected constructor(protected m_score1: number, protected m_score2: number) {
    }

    public isCurrentScore(): boolean {
        throw new Error('Not Implemented')
    }

    public render(): string {
        throw new Error('Not Implemented')
    }

    protected scoreDifference() {
        return this.m_score1 - this.m_score2;
    }

    protected playerInAdvantage() {
        if (this.m_score1 > this.m_score2)
            return "player1"
        if (this.m_score2 > this.m_score1)
            return "player2"
        return null
    }
}

export class InProgressScore extends Score {
    constructor(m_score1: number, m_score2: number) {
        super(m_score1, m_score2);
    }

    public isCurrentScore(): boolean {
        return (this.m_score1 !== this.m_score2) && (this.m_score1 < 4 && this.m_score2 < 4);
    }

    public render(): string {
        let tempScore: number = 0;
        let score = ''
        for (let i = 1; i < 3; i++) {
            if (i === 1) tempScore = this.m_score1;
            else {
                score += '-';
                tempScore = this.m_score2;
            }
            switch (tempScore) {
                case 0:
                    score += 'Love';
                    break;
                case 1:
                    score += 'Fifteen';
                    break;
                case 2:
                    score += 'Thirty';
                    break;
                case 3:
                    score += 'Forty';
                    break;
            }
        }
        return score;
    }
}

export class TieScore extends Score {
    constructor(m_score1: number, m_score2: number) {
        super(m_score1, m_score2);
    }

    public isCurrentScore(): boolean {
        return this.m_score1 === this.m_score2;
    }

    public render(): string {
        switch (this.m_score1) {
            case 0:
                return 'Love-All';
            case 1:
                return 'Fifteen-All';
            case 2:
                return 'Thirty-All';
            default:
                return 'Deuce';
        }
    }
}

export class AdvantageScore extends Score {
    constructor(m_score1: number, m_score2: number) {
        super(m_score1, m_score2);
    }

    public isCurrentScore(): boolean {
        return (this.m_score1 >= 4 || this.m_score2 >= 4) && (Math.abs(this.scoreDifference()) === 1);
    }

    public render(): string {
        return `Advantage ${this.playerInAdvantage()}`;

    }
}

export class EndScore extends Score {
    constructor(m_score1: number, m_score2: number) {
        super(m_score1, m_score2);
    }

    public isCurrentScore(): boolean {
        return (this.m_score1 >= 4 || this.m_score2 >= 4) && (Math.abs(this.scoreDifference()) > 1);
    }

    public render(): string {
        if (this.scoreDifference() >= 2) return 'Win for player1';
        return 'Win for player2';
    }

}
